### DISCUSSION TITLE
Why does HashiCorp vault API URL have '/config' in it?

### DISCUSSION THREAD

**Discussion by garywhiteford** [apache/airflow#24603] [Q&A]
https://github.com/apache/airflow/blob/0527a0b6ce506434a23bc2a6f5ddb11f492fc614/airflow/providers/hashicorp/secrets/vault.py#L242

This results in an error in the API call.

\<Screenshot>
![2022-06-22_8-14-56](https://user-images.githubusercontent.com/33525443/175042066-a2ef69dd-8489-4b75-b2ae-1d42ab573063.png)

**Comment by potiuk** [comment]
Maybe because of that? I am guessing but I actually read the docstring and this is what it says:

```
    :param config_path: Specifies the path of the secret to read Airflow Configurations
        (default: 'config'). If set to None (null), requests for configurations will not be sent to Vault.
```

**Comment by garywhiteford** [comment]
Please bear with me.

So, in ..\airflow\configuration.py, if there is an environment variable with a suffix of _SECRET, it loads from the environment variable that value (a string with **_pathToTarget/targetName_** that we have stored as such in our HashiCorp vault) 
![image](https://user-images.githubusercontent.com/33525443/175127093-3e3bc310-a864-4976-b706-4deb2633d338.png)
Note: Line 459 in my screenshot above is now line 463 (last changed 2 years ago).
https://github.com/apache/airflow/blob/c3c1f7ea2851377ba913075844a8dde9bfe6376e/airflow/configuration.py#L463

… and elsewhere in ..\airflow\configuration.py, it sets the secrets_client and calls its **get_config** _method_ (forgive any mistakes in nomenclature)
![image](https://user-images.githubusercontent.com/33525443/175127343-75a167e0-b35c-4370-8a0b-ae2b7059fffc.png)
https://github.com/apache/airflow/blob/c3c1f7ea2851377ba913075844a8dde9bfe6376e/airflow/configuration.py#L120

… which takes the target name (as key) and, for reasons that are not entirely clear to me, pre-pends a default **config_path** to it
![image](https://user-images.githubusercontent.com/33525443/175127419-71e17ba7-af72-499e-975e-0fdd76dbdd8f.png)
https://github.com/apache/airflow/blob/c3c1f7ea2851377ba913075844a8dde9bfe6376e/airflow/providers/hashicorp/secrets/vault.py#L242

… which default **config_path**, as pointed out, is set to be ‘config’ 
![image](https://user-images.githubusercontent.com/33525443/175127518-c9e5ea2d-0ae8-4eef-9ccf-671d45731f42.png)
https://github.com/apache/airflow/blob/c3c1f7ea2851377ba913075844a8dde9bfe6376e/airflow/providers/hashicorp/secrets/vault.py#L51-L52

… and results in a now-incorrect **secret_path** [what's the markdown for red?]
![image](https://user-images.githubusercontent.com/33525443/175127639-06be052b-a2d0-4578-bef1-fb889af48968.png)

### Key issue
… which incorrect **secret_path** [in red text], in HashiCorp vault (..\airflow\providers\hashicorp\_internal_client\vault_client.py), is used as an argument to the HashiCorp standard **read_secret_version** function, 
![image](https://user-images.githubusercontent.com/33525443/175127743-4fec68b5-268f-4bd1-8755-aa58212a0b30.png)
https://github.com/apache/airflow/blob/c3c1f7ea2851377ba913075844a8dde9bfe6376e/airflow/providers/hashicorp/_internal_client/vault_client.py#L347-L349

… which results in an incorrect URL for the API call as shown in the original screenshot (see https://github.com/apache/airflow/discussions/24603#discussion-4165047 above).

Has no one run into this before?

**Comment by garywhiteford** [comment]
Also, as I am stepping through the same commands in Python with the hvac package, I see that I can properly return the contents of a path/folder (as an array, I guess), but not an individaul item in the path/folder. 

\<Screenshot 1\>
![image](https://user-images.githubusercontent.com/33525443/175144165-746b1385-fb4f-472e-b5af-ea64c61490c9.png)


So, I would guess that the backend needs to know that it is obtaining a path and then pick the individual element out of the array based on the value of the key/value pair from the environment variable. 

It looks like someone was 2/3 of the way there with ['data']['data'], they just need to add ['data']['data']['key'] (I'm guessing -- I'm not a Python expert). 

\<Screenshot 2\>
![image](https://user-images.githubusercontent.com/33525443/175144564-7150ab37-18a1-4f47-a812-2c258d30f072.png)
https://github.com/apache/airflow/blob/c3c1f7ea2851377ba913075844a8dde9bfe6376e/airflow/providers/hashicorp/_internal_client/vault_client.py#L354

Has this been identified anywhere?

**Reply by potiuk** [reply]
No idea. Airflow has 73 providers an some ~5000 integration classes. We merge 300 Prs a month and there are 300 issues opend a moth.  It's likely there are some rare path that have some errors (like every software :)) 

Hard to answer your questions - or actually, it's easy - just search through the issues. There is no-one who keeps track of all issues, but - luckily those issues are public and searchable. So feel free to check it (you have precisely the same capabilities like everyone here - including maintainers). 

I do not know enough details of Hashicorp to dicuss the screenshots. But Airflow is created by > 2000 contributors like you - if you can fix it locally and provide a Pull Request fixing it (with unit tests) then the discsussion might be more productive. Even if it will be rejected, everyone will learn from discussion there - looking at the actual code rather than screenshots (even if you made a lot of effort on making pointy screenshots, actually looking at modified code and discussing it there is what Pull Request interface has been created for.

You seem to be good enough  and have all the tools to understand and debug the code, and you can actually test it with real Hashicorp - so proably you are one of the best people in the world to attempt to fix it. Just follow our CONTRIBUTING.rst to do so.

I encourage you to do so to make the discussion more productive.

**Comment by garywhiteford** [comment]
Thank you, Jarek! (I worked with a Jarek at my last job.) I'm very much inclined to contribute. I need to get up to speed on the coding and unit testing piece. Right now I'm a questionable "Monday Morning Quarterback." I did try searching multiple ways but did not find any other discussions. We actually have an outside team working with us this summer. Perhaps they can contribute to this as part of their project.

**Reply by potiuk** [reply]
Cool!
